<launch>
    <arg name="use_sim_time" default="true"/>
    <param name="/use_sim_time" value="true"/>

    <node name="orb_slam2_stereo" pkg="ORB_SLAM2" type="Stereo" output="screen"
    	args="/root/catkin_ws/src/ORB_SLAM2-1.0.0/Vocabulary/ORBvoc.txt 
               /root/catkin_ws/src/ORB_SLAM2-1.0.0/Examples/Stereo/NUFR_RECTIFIED.yaml 
               false">
        
        <remap from="camera/left/image_rect"   to="/zed2i/zed_node/left/image_rect_color"/>
        <remap from="camera/right/image_rect"  to="/zed2i/zed_node/right/image_rect_color"/>   
        <remap from="camera/left/camera_info"  to="/zed2i/zed_node/left/camera_info"/>
        <remap from="camera/right/camera_info" to="/zed2i/zed_node/right/camera_info"/> 
        
        <param name="map_frame_id" value="map"/>
        <param name="base_frame_id" value="zed_left_camera_frame"/>
        <param name="use_sim_time" value="$(arg use_sim_time)"/>
    </node>

    <node name="elas" pkg="elas_ros" type="elas_ros" output="screen">
        <remap from="stereo/left/image"        to="/zed2i/zed_node/left/image_rect_color"/>
        <remap from="stereo/right/image"       to="/zed2i/zed_node/right/image_rect_color"/>   
        <remap from="stereo/left/camera_info"  to="/zed2i/zed_node/left/camera_info"/>
        <remap from="stereo/right/camera_info" to="/zed2i/zed_node/right/camera_info"/>
        
        <param name="initial_variance" type="double" value="3.0"/>
        <param name="measurement_variance" type="double" value="0.5"/>
        <param name="variance_threshold" type="double" value="1.0"/>
        <param name="stability_threshold" type="int" value="0"/>

        <param name="disp_min" type="int" value="0"/>
        <param name="disp_max" type="int" value="255"/>
        <param name="support_threshold" type="double" value="0.80"/>
        <param name="support_texture" type="int" value="10"/>
        <param name="candidate_stepsize" type="int" value="10"/>
        <param name="incon_window_size" type="int" value="5"/>
        <param name="incon_threshold" type="int" value="5"/>
        <param name="incon_min_support" type="int" value="5"/>
        <param name="add_corners" type="bool" value="1"/>
        <param name="grid_size" type="int" value="20"/>
        <param name="beta" type="double" value="0.02"/>
        <param name="gamma" type="double" value="3"/>
        <param name="sigma" type="double" value="1"/>
        <param name="sradius" type="double" value="2"/>
        <param name="match_texture" type="int" value="1"/>
        <param name="lr_threshold" type="int" value="2"/>
        <param name="speckle_sim_threshold" type="double" value="1"/>
        <param name="speckle_size" type="int" value="200"/>
        <param name="ipol_gap_width" type="int" value="50"/>
        <param name="filter_median" type="bool" value="0"/>
        <param name="filter_adaptive_mean" type="bool" value="1"/>
        <param name="postprocess_only_left" type="bool" value="1"/>
        <param name="subsampling" type="bool" value="0"/>
        
        <param name="approximate_sync" value="true" />
        <param name="queue_size" type="int" value="10"/>
        <param name="use_temporal_filter" type="bool" value="true"/>
    </node>


    <node pkg="pointcloud_mapping" type="mapping" name="map_builder" output="screen">
        <param name="use_sim_time" value="$(arg use_sim_time)"/>
        <param name="world_frame" value="map"/>
        <param name="topic_pointcloud" type="string" value="/elas/point_cloud" />
        <param name="resolution" type="double" value="0.1"/>
        <param name="max_range" type="double" value="7"/>
        <param name="fusion_radius" type="double" value="0.04"/>
        <param name="normal_similarity_threshold" type="double" value="0.85"/>
        <param name="min_update_count_for_keep" type="int" value="2"/>
        <param name="clean_map_interval" type="int" value="10"/> 
    </node>

    <!-- <node name="voxblox_node" pkg="voxblox_ros" type="tsdf_server" output="screen">
        <remap from="pointcloud" to="/elas/point_cloud"/>
        <param name="world_frame" value="map"/>
        <param name="use_sim_time" value="$(arg use_sim_time)"/>
        <param name="method" value="simple" />
        <param name="tsdf_voxel_size" value="0.01" />
        <param name="tsdf_truncation_distance" value="0.05" />
        <param name="pointcloud_queue_size" value="20" />
        <param name="publish_tsdf_map" value="true" />
        <param name="publish_mesh" value="true" />
        <param name="mesh_filename" value="/root/catkin_ws/output/mapping/final_tsdf_map.ply" />
    </node> -->

    <node pkg="rviz" type="rviz" name="rviz" respawn="false" output="screen"
         args="-d $(find mapping)/rviz/NUFR_M3F.rviz"> 
    </node>

</launch>
